### PR TITLE
Partner Portal: Add persistence to Partner and Partner Keys

### PR DESCRIPTION
**Pull request by IanRamosC**
#### Changes proposed in this Pull Request

Context: 1187494150150258-as-1200070129781595/f

* This PR adds `withSchemaValidation` e `withPersistence` to the `activePartnerKeyReducer` and `currentPartnerReducer` to prevent the Partner Key selection screen when refreshing or changing the page.

#### Testing instructions

* If you do not have a partner key, please contact Infinity for one or you will be unable to test this PR.
* Check out this PR locally, then run* `yarn && ENABLE_FEATURES=no-force-sympathy && yarn start-jetpack-cloud`.
* Visit http://jetpack.cloud.localhost:3000/partner-portal
* Select a Partner Key
* Refresh the page
* Make sure it doesn't redirect you to the "Select a Partner Key" screen

Note: Calypso is likely to skip the initial state rehydration as mentioned [here](https://github.com/Automattic/wp-calypso/pull/51317#issuecomment-805036236), and this PR might not work correctly. To avoid that, we have to enable the  `no-force-sympathy` by passing `ENABLE_FEATURES=no-force-sympathy` before running `yarn start-jetpack-cloud`.

### CODE REVIEW

**Comment by poisa** [comment]
This is working sporadically for me. In this video the only actions I did was refresh and select a key and it doesn't always work.

https://user-images.githubusercontent.com/348248/112179240-81109600-8bc8-11eb-8933-23b890aebd1a.mov

**Comment by atanas-dev** [comment]
@poisa in the local development environment Calypso will randomly throw out the cached storage for testing purposes. Give it another try and see if the "failures" coincide with this message in the console:
Screenshot 2021-03-23 at 18 15 35
(I haven't tested the PR yet but this looks very likely to be the cause)

**Comment by poisa** [comment]
> @poisa in the local development environment Calypso will randomly throw out the cached storage for testing purposes. Give it another try and see if the "failures" coincide with this message in the console:

Oh wow, what a head scratcher. Yes, this is exactly the case. Probably a good idea to include this in the testing instructions. Is there a way to disable this?

**Comment by IanRamosC** [comment]
> Oh wow, what a head scratcher. Yes, this is exactly the case. Probably a good idea to include this in the testing instructions. Is there a way to disable this?

Thanks @poisa and @atanas-dev. I'm going to update the PR description to include this.

File: client/state/partner-portal/partner/schema.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
+export const activePartnerKeySchema = {
+	type: 'number',
+};
+
+export const currentPartnerSchema = {
+	type: 'object',
+	additionalProperties: true,
+	required: [ 'id', 'slug' ],
```
Comment (atanas-dev) [inline client/state/partner-portal/partner/schema.ts]:
```suggestion
	required: [ 'id', 'slug', 'keys' ],
```

File: client/state/partner-portal/partner/schema.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+	patternProperties: {
+		'^\\d+$': {
+			id: { type: 'number' },
+			name: { type: 'string' },
+			slug: { type: 'string' },
+			keys: {
+				type: [ 'array', 'null' ],
```
Comment (atanas-dev) [inline client/state/partner-portal/partner/schema.ts]: Can keys be null? If it can then IMO we should fix it so it returns an empty array and not null.

File: client/state/partner-portal/partner/schema.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+				items: {
+					type: 'object',
+					required: [ 'id', 'oAuthToken' ],
```
Comment (atanas-dev) [inline client/state/partner-portal/partner/schema.ts]:
```suggestion
					required: [ 'id', 'oAuthToken', 'name' ],
```

File: client/state/partner-portal/partner/reducer.ts [hunk 4]
```diff
@@ -72,7 +81,12 @@ export const current = withoutPersistence( ( state = initialState.current, actio
 	}
 
 	return state;
-} );
+};
+
+export const current = withSchemaValidation(
+	currentPartnerSchema,
+	withPersistence( currentPartnerReducer )
+);
```
Comment (atanas-dev) [inline client/state/partner-portal/partner/reducer.ts]: We should not be caching the partner and the keys:
1. AFAIK cache goes in localStorage which is NOT secure and should not be used for secrets.
2. If a partner chooses a key (which gets caches), then an a11n disables that key, then the user refreshes the page it will attempt to use the cached key (which no longer exists as it won't be returned by the API) resulting in failing requests and an empty key selection dropdown
Screenshot 2021-03-24 at 18 13 31

You need 3 keys to test this as if you have 2 and deactivate one the auto key selection will kick in and hide the bug.

Note: You need to test with 3 CONSECUTIVE active keys as otherwise you will hit this bug: 1187494150150258-as-1200106791469029

---

In summary, we only want to cache the selected key id, not the partner and keys themselves. When the partner and its keys are loaded from the API, we validate the key exists and auto select it.

**Review by atanas-dev — CHANGES_REQUESTED** [review]

File: client/state/partner-portal/partner/actions.ts [hunk 5]
```diff
@@ -51,17 +52,29 @@ export function fetchPartner( dispatch: ReduxDispatch, getState: () => PartnerPo
...
+		if ( ! isPartnerKeyAvailable ) {
+			// do something else
+			return;
+		}
 
-		if ( keys.length === 1 ) {
+		if ( numberOfKeys === 1 ) {
 			dispatch( setActivePartnerKey( keys[ 0 ] ) );
+		} else if ( numberOfKeys > 1 ) {
+			dispatch( setActivePartnerKey( currentPartnerKeyId ) );
```
Comment (atanas-dev) [inline client/state/partner-portal/partner/actions.ts]: I think replacing the logic with this can make it easier to understand what is happening:
```js
const activePartnerKeyId = getActivePartnerKeyId( getState() );
const keys = partner.keys.map( ( key ) => key.id );
let newKeyId = 0;

if ( keys.length === 1 ) {
	// If we only get a single key, auto-select it for convenience.
	newKeyId = keys[ 0 ];
}

if ( keys.includes( activePartnerKeyId ) ) {
	// If the active key id is for a valid key, select it.
	newKeyId = activePartnerKeyId;
}

if ( newKeyId ) {
	dispatch( setActivePartnerKey( newKeyId ) );
}
```
Also it will cover the situation where a cached key is no longer available and the user has only one available key now but they will still be asked to choose it because of this early return:
```
if ( ! isPartnerKeyAvailable ) {
	// do something else
	return;
}
```

**Review by atanas-dev — CHANGES_REQUESTED** [review]
This has the same problem when you deactivate a key that has been cached:
Screenshot 2021-04-15 at 15 31 34
This is fixable by reverting the changes to controller.tsx as getActivePartnerKey validates that the selected key exists and if it doesn't then the controller will ask the user to select a key that exists.

File: client/state/partner-portal/partner/actions.ts [hunk 6]
```diff
@@ -51,17 +52,29 @@ export function fetchPartner( dispatch: ReduxDispatch, getState: () => PartnerPo
...
 export function receivePartner( partner: Partner ): PartnerPortalThunkAction {
-	return ( dispatch: ReduxDispatch ): void => {
+	return ( dispatch: ReduxDispatch, getState: () => PartnerPortalStore ): void => {
 		dispatch( {
 			type: JETPACK_PARTNER_PORTAL_PARTNER_RECEIVE,
 			partner,
 		} );
 
-		// If we only get a single key, auto-select it for the user for simplicity.
+		// We check the active key otherwise.
```
Comment (atanas-dev) [inline client/state/partner-portal/partner/actions.ts]: Leftover comment from a previous revision.

**Review by atanas-dev — APPROVED** [review]
Tests good for me 👍
